Multiple file uploads with more than one part

### Expected Behavior

Uploading multiple files with different content types should work.

### Actual Behaviour

Trying to upload files in MultiPartBody results in errors.

### Steps To Reproduce

Create a controller with a method
```
@Post(value = "/{+path}", consumes = MediaType.MULTIPART_FORM_DATA, produces = MediaType.APPLICATION_JSON)
public Mono<List<String>> upload(@PathVariable String path,
                                     @Part("file") Publisher<CompletedFileUpload> file,
                                     @Nullable @Part("metadata") Metadata metadata)
return Flux.from(file)
                .subscribeOn(Schedulers.boundedElastic())
                .map(part -> { ... process part ...; return part.getFilename(); })
                .collectList();
```

Then try to upload one or more files:
`curl -vF 'file=@./image1.jpeg' -F 'file=@./image2.jpeg' http://localhost:8080/uploads`

{"_links":{"self":[{"href":"/uploads","templated":false}]},"_embedded":{"errors":[{"message":"Internal Server Error: refCnt: 0, decrement: 1"}]},"message":"Internal Server Error"}%

Uploading multiple files gives another error:

`curl -vF 'file=@./image1.jpeg' -F 'file=@./image2.jpeg' -F 'metadata=@./metadata.json' http://localhost:8080/uploads`

{"_links":{"self":[{"href":"/uploads","templated":false}]},"_embedded":{"errors":[{"message":"Internal Server Error: The mapper [io.micronaut.http.server.netty.binders.NettyPartUploadAnnotationBinder$$Lambda/0x00000008015b9fd0] returned a null value."}]},"message":"Internal Server Error"}%   


### Environment Information

- Operating System: MacOS 14.3.1
- JDK Version: 21.0.2

### Example Application

https://github.com/goeh/micronaut-multiple-file-upload-issue

### Version

4.3.4

If I change the method signature to `Flux<String>` and remove `.collectList()` I get another error.
From the logging I can see that the parts are actually uploaded, but then `IllegalReferenceCountException` is thrown.
```
10:05:56.081 [boundedElastic-1] INFO  micronaut.issue.UploadController - IMG_0726.jpeg uploaded to uploads
10:05:56.082 [boundedElastic-1] INFO  micronaut.issue.UploadController - IMG_0726.jpeg uploaded to uploads
10:05:56.083 [default-nioEventLoopGroup-1-3] WARN  i.m.h.s.n.h.PipeliningServerHandler - Reactive response received an error after some data has already been written. This error cannot be forwarded to the client.
io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:83)
	at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:148)
	at io.netty.util.AbstractReferenceCounted.release(AbstractReferenceCounted.java:76)
	at io.micronaut.http.server.netty.multipart.NettyCompletedFileUpload.getInputStream(NettyCompletedFileUpload.java:89)
	at micronaut.issue.UploadController.lambda$upload$0(UploadController.java:32)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:106)
	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.onNext(FluxSubscribeOn.java:151)
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:194)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onNext(FluxPeekFuseable.java:503)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onNext(FluxPeekFuseable.java:503)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onNext(FluxPeekFuseable.java:503)
	at reactor.core.publisher.FluxContextWriteRestoringThreadLocals$ContextWriteRestoringThreadLocalsSubscriber.onNext(FluxContextWriteRestoringThreadLocals.java:118)
	at reactor.core.publisher.FluxContextWriteRestoringThreadLocals$FuseableContextWriteRestoringThreadLocalsSubscriber.onNext(FluxContextWriteRestoringThreadLocals.java:190)
	at reactor.core.publisher.SinkManyUnicast.drainRegular(SinkManyUnicast.java:284)
	at reactor.core.publisher.SinkManyUnicast.drain(SinkManyUnicast.java:365)
	at reactor.core.publisher.SinkManyUnicast.request(SinkManyUnicast.java:434)
	at reactor.core.publisher.FluxContextWriteRestoringThreadLocals$ContextWriteRestoringThreadLocalsSubscriber.request(FluxContextWriteRestoringThreadLocals.java:163)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.request(FluxPeekFuseable.java:437)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.request(FluxPeekFuseable.java:437)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.request(FluxPeekFuseable.java:437)
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.request(FluxHandleFuseable.java:260)
	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.lambda$requestUpstream$0(FluxSubscribeOn.java:135)
	at io.micrometer.context.ContextSnapshot.lambda$wrap$0(ContextSnapshot.java:91)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
10:05:56.083 [default-nioEventLoopGroup-1-3] WARN  i.n.u.c.AbstractEventExecutor - A task raised an exception. Task: io.micronaut.http.server.netty.handler.PipeliningServerHandler$StreamingOutboundHandler$$Lambda/0x00000070015c9880@450aa085
io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:83)
	at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:148)
	at io.netty.util.AbstractReferenceCounted.release(AbstractReferenceCounted.java:76)
	at io.micronaut.http.server.netty.FormRouteCompleter.release(FormRouteCompleter.java:234)
	at io.micronaut.http.server.netty.body.ManagedBody.release(ManagedBody.java:102)
	at io.micronaut.http.server.netty.body.ManagedBody.release(ManagedBody.java:102)
	at io.micronaut.http.server.netty.NettyHttpRequest.release(NettyHttpRequest.java:439)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.cleanupRequest(RoutingInBoundHandler.java:153)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.responseWritten(RoutingInBoundHandler.java:170)
	at io.micronaut.http.server.netty.websocket.NettyServerWebSocketUpgradeHandler.responseWritten(NettyServerWebSocketUpgradeHandler.java:167)
	at io.micronaut.http.server.netty.handler.PipeliningServerHandler$StreamingOutboundHandler.onError(PipeliningServerHandler.java:1131)
	at io.micronaut.http.server.netty.handler.PipeliningServerHandler$StreamingOutboundHandler.lambda$onError$1(PipeliningServerHandler.java:1121)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
10:05:56.083 [default-nioEventLoopGroup-1-3] WARN  i.n.channel.DefaultChannelPipeline - An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
io.netty.channel.ChannelPipelineException: io.micronaut.http.server.netty.handler.PipeliningServerHandler.handlerRemoved() has thrown an exception.
	at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:640)
	at io.netty.channel.DefaultChannelPipeline.destroyDown(DefaultChannelPipeline.java:876)
	at io.netty.channel.DefaultChannelPipeline.destroyUp(DefaultChannelPipeline.java:844)
	at io.netty.channel.DefaultChannelPipeline.destroy(DefaultChannelPipeline.java:836)
	at io.netty.channel.DefaultChannelPipeline.access$700(DefaultChannelPipeline.java:46)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelUnregistered(DefaultChannelPipeline.java:1392)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:215)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:195)
	at io.netty.channel.DefaultChannelPipeline.fireChannelUnregistered(DefaultChannelPipeline.java:821)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:821)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:83)
	at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:148)
	at io.netty.util.AbstractReferenceCounted.release(AbstractReferenceCounted.java:76)
	at io.micronaut.http.server.netty.FormRouteCompleter.release(FormRouteCompleter.java:234)
	at io.micronaut.http.server.netty.body.ManagedBody.release(ManagedBody.java:102)
	at io.micronaut.http.server.netty.body.ManagedBody.release(ManagedBody.java:102)
	at io.micronaut.http.server.netty.NettyHttpRequest.release(NettyHttpRequest.java:439)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.cleanupRequest(RoutingInBoundHandler.java:153)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.responseWritten(RoutingInBoundHandler.java:170)
	at io.micronaut.http.server.netty.websocket.NettyServerWebSocketUpgradeHandler.responseWritten(NettyServerWebSocketUpgradeHandler.java:167)
	at io.micronaut.http.server.netty.handler.PipeliningServerHandler$StreamingOutboundHandler.discard(PipeliningServerHandler.java:1172)
	at io.micronaut.http.server.netty.handler.PipeliningServerHandler.handlerRemoved(PipeliningServerHandler.java:199)
	at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:1138)
	at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
	... 17 common frames omitted
```

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/binders/NettyPartUploadAnnotationBinder.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/binders/NettyPartUploadAnnotationBinder.java
@@ -96,7 +96,13 @@ final class NettyPartUploadAnnotationBinder<T> implements AnnotatedRequestArgume
 
             @Override
             public Optional<T> getValue() {
-                return completableFuture.getNow(Optional.empty());
+                Optional<T> res = completableFuture.getNow(Optional.empty());
+                //noinspection OptionalAssignedToNull
+                if (res == null) {
+                    // tricky: If the Mono completes without an element, the future will return null here.
+                    res = Optional.empty();
+                }
+                return res;
             }
         };
     }

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/binders/NettyPartUploadAnnotationBinder.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/binders/NettyPartUploadAnnotationBinder.java
@@ -78,8 +78,8 @@ final class NettyPartUploadAnnotationBinder<T> implements AnnotatedRequestArgume
         if (skipClaimed && nettyRequest.formRouteCompleter().isClaimed(inputName)) {
             return BindingResult.unsatisfied();
         }
-        CompletableFuture<T> completableFuture = Mono.from(nettyRequest.formRouteCompleter().claimFieldsComplete(inputName))
-            .map(d -> NettyConverters.refCountAwareConvert(conversionService, d, context).orElse(null))
+        CompletableFuture<Optional<T>> completableFuture = Mono.from(nettyRequest.formRouteCompleter().claimFieldsComplete(inputName))
+            .map(d -> NettyConverters.refCountAwareConvert(conversionService, d, context))
             .toFuture();
 
         return new PendingRequestBindingResult<>() {
@@ -96,7 +96,7 @@ final class NettyPartUploadAnnotationBinder<T> implements AnnotatedRequestArgume
 
             @Override
             public Optional<T> getValue() {
-                return Optional.ofNullable(completableFuture.getNow(null));
+                return completableFuture.getNow(Optional.empty());
             }
         };
     }

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/FileUploadSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/FileUploadSpec.groovy
@@ -2,9 +2,11 @@ package io.micronaut.http.server.netty
 
 import io.micronaut.context.ApplicationContext
 import io.micronaut.context.annotation.Requires
+import io.micronaut.core.annotation.Nullable
 import io.micronaut.http.HttpRequest
 import io.micronaut.http.MediaType
 import io.micronaut.http.annotation.Controller
+import io.micronaut.http.annotation.Part
 import io.micronaut.http.annotation.Post
 import io.micronaut.http.annotation.Produces
 import io.micronaut.http.client.HttpClient
@@ -34,6 +36,26 @@ class FileUploadSpec extends Specification {
         server.stop()
     }
 
+    def 'wrong content type'() {
+        given:
+        def ctx = ApplicationContext.run(['spec.name': 'FileUploadSpec'])
+        def server = ctx.getBean(EmbeddedServer)
+        server.start()
+        def client = ctx.createBean(HttpClient, server.URI).toBlocking()
+
+        when:
+        def response = client.retrieve(HttpRequest.POST(
+                '/multipart/deserialize',
+                MultipartBody.builder().addPart('metadata', 'metadata', MediaType.APPLICATION_OCTET_STREAM_TYPE, '{"foo":"bar"}'.bytes).build())
+                .contentType(MediaType.MULTIPART_FORM_DATA_TYPE), String)
+        then:
+        response == 'Metadata: null'
+
+        cleanup:
+        client.close()
+        server.stop()
+    }
+
     @Controller('/multipart')
     @Requires(property = 'spec.name', value = 'FileUploadSpec')
     @Produces(MediaType.TEXT_PLAIN)
@@ -43,5 +65,12 @@ class FileUploadSpec extends Specification {
             def bytes = data.inputStream.bytes
             return "Uploaded ${bytes.length} bytes"
         }
+
+        @Post(value = '/deserialize', consumes = MediaType.MULTIPART_FORM_DATA)
+        String completeFileUpload(@Nullable @Part("metadata") Metadata metadata) {
+            return "Metadata: " + metadata
+        }
     }
+
+    record Metadata(@Nullable String foo) {}
 }

